Web3 Module API Documentation

nivida opened the pull request:
## Description

Adding of a detailed Web3 Module API Documentation.

Steps before this documentation can be merged: 

- [x] Document all ES6 classes regarding the Web3 Module API 
- [x] Link the classes to the source code until a class reference exist.

## Type of change

- [x] Documentation

Commit:
web3-modules documentation started

Commit:
Merge branch '1.0' into documentation/web3-modules

Commit:
web3-modules.rst updated

Commit:
AbstractMethod, and AbstractMethodFactory added to module documentation

Commit:
header updated

Commit:
module api docs splitted and AbstractCallMethod documented

Commit:
All classes documented and example section added

Commit:
typos and formatting

Commit:
typos

Commit:
example fixed

Commit:
section added to show the already implemented methods and subscriptions classes a devloper could extend from

Commit:
typo

Commit:
AbstractSubscription added to module api documentation

Commit:
rst titles fixed

Commit:
missing break added

Commit:
breaks removed

Commit:
include file renamed

Levino commented:
Hm..

nivida commented:
@Levino If you find the time could you please check this documenation if it's understandable for you?

levino commented:
First feedback: I would create an issue for this and reference here with "Fixes #1928282828".
Please update with base branch.

I will continue to review.

levino commented:
- Build does not pass.
- Why is there no preview link? How are the docs hosted?

levino commented:
There is no documentation on how to build the docs locally. There is no hot watcher is there? Created https://github.com/ethereum/web3.js/issues/2250

Levino commented:
Created https://github.com/ethereum/web3.js/issues/2250

levino commented:
How do you want me to review this? Checkout the branch, run `npm run docs` and then go through the PDFs? If https://github.com/ethereum/web3.js/issues/2251 was implemented this would be much nicer.

levino commented on packages/web3-core-method/lib/methods/AbstractMethod.js:
```diff
@@ -212,7 +212,7 @@ export default class AbstractMethod {
      */
     isHash(parameter) {
-        return isString(parameter) && parameter.indexOf('0x') === 0;
+        return isString(parameter) && parameter.startsWith('0x') === 0;
```
I do not see a test changed for this. This is troubling. Either there is no test or the test does not test anything. How could it have worked before? Also I consider this a mix of concerns: I thought this PR is about the docs, now I see code changes. I think the code changes should be moved to another PR.

levino commented:
I cannot build this locally.

```bash
npm run docs

> web3@1.0.0-beta.38 docs /home/levin/coding/hardfork/web3.js
> cd docs; make html;

sphinx-build -b html -d _build/doctrees   . _build/html
make: sphinx-build: Command not found
Makefile:53: recipe for target 'html' failed
make: *** [html] Error 127
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! web3@1.0.0-beta.38 docs: `cd docs; make html;`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the web3@1.0.0-beta.38 docs script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/levin/.npm/_logs/2019-01-28T11_28_10_130Z-debug.log
```

So one needs to globally install `sphinx-build` on the local machine? Kind of a bad setup. Is it not possible to install this via npm?

levino commented:
So I will be waiting for you to get back on my comments / questions. Please at least add a "how to run locally".

nivida commented:
Thanks for your feedback. Yes, I know the documentation setup is not good. The goal of this PR is to add an additional section to the documentation, any other improvements will be done later. The focus currently is on bugs and to add the Web3 Moudle API documentation.

You do not have to checkout the branch for reviewing the content of the documentation. It's also possible to view them directly on [github](https://github.com/ethereum/web3.js/blob/95087d6b1977f9ff07824d083b3c457f00ec00b0/docs/include_web3-module-abstract-method-class-reference.rst).

levino commented:
So it is just markdown.

levino commented:
I think it would be much easier if you created one PR for each point in your list above. Having so many issues in one PR is very confusing.

levino commented:
I do not understand what the point of this PR should be. Could you please verbosely explain the issue that you are trying to solve?

Before merging this PR: Who wants to do what?

Then I can see whether or not that person achieves what they wanted to do after this PR has been merged.

nivida commented:
I only asked if you could read the documentation and tell me if you understand the content of it without any additional information. The point of this PR is to document the Web3 Module API as the title already says. 

If you find the time could you quickly read through the document and tell me if you are able to create a module with the provided information or not?

Commit:
web3 module documenation updated

levino commented:
What is a module and why would I create one? I really cannot follow. It is very hard because this is not embedded in the docs atm and I lack context.

OFRBG commented:
@Levino @nivida I'd like to chime in. I'm highly interested in making web3js documentation as readable as possible. Would it be okay if I propose a specific structure for each topic? I've been using the docs for almost two years now, and I'm relatively familiar with the problems it has.

levino commented:
Please start by creating issues that describe the problems you encounter in a very understandable (I am a little slow it seems so please consider this) manner. If everyone agrees that these issues are worth fixing, you can start on the fixes.

Commit:
web3-modules.rst file restructured

Commit:
web3-module added in index.rst

yxliang01 commented:
I would say the `extend` method should have been retained until this documentation is ready... This transition makes life of developers extending web3 difficult... Maybe it's even better if the Web3 doc allows us to view the doc of older 1.0 beta versions.

Commit:
web3-module.rst splitted up

fixanoid commented:
> I would say the `extend` method should have been retained until this documentation is ready... This transition makes life of developers extending web3 difficult... Maybe it's even better if the Web3 doc allows us to view the doc of older 1.0 beta versions.

And I would add that removing this breaks a ton of modules already in use. And while this is helpful in the longer run, it leaves many a module broken until maintainers update them or manual dependency resolution is made to the last working version. Why not leave it available and throw warnings for a while?

Commit:
web3-module-contract.rst file added

nivida commented:
@fixanoid 
> And I would add that removing this breaks a ton of modules already in use

I've removed this possibility now a month ago and explained why I removed this. Feel free to create a PR if there is a missing JSON-RPC method.

> And while this is helpful in the longer run, it leaves many a module broken until maintainers update them or manual dependency resolution is made to the last working version. 

If you would like to use the ``extend`` method then please use the version beta.37.

> Why not leave it available and throw warnings for a while?

I didn't add this during the refactoring because it's really ugly and hard to test. There are also some problems with the formatters and utils modules which the ``extend`` method was indirectly depending on. 

All the details about the refactoring are explained in the release announcement of version beta.38: 
https://medium.com/@samuel_91690/web3-js-v1-0-0-beta-38-8d2bb6e73d0b

Commit:
Merge branch '1.0' into documentation/web3-modules

Commit:
web3-core.rst updated

Commit:
web3-modules documenation updated

Commit:
web3-core-methods.rst updated

Commit:
web3-core-method and web3-core-subscriptions documentation upated

Commit:
include_web3-module-abstract-method-class-reference.rst updated

Commit:
references fixed for the AbstractMethod documentation

nivida committed:
Merge branch '1.0' into documentation/web3-modules

Commit:
web3-module api documentation updated

Commit:
index.rst updated

Commit:
intro updated in web3-module documentation

Commit:
type property documentation updated

Commit:
code blocks added to the web3-module documentation

Commit:
titles fixed in docs/include_web3-module-abstract-method-class-reference.rst

Commit:
method arguments setter/getter documentation updated

Commit:
docs/include_web3-module-abstract-method-class-reference.rst updated

Commit:
not required breaks removed in the web3-module documentation

Commit:
docs/_build folder removed and added to gitignore

Commit:
source code url updated in web3-core.rst

Commit:
web3-module documentation improved

nivida committed:
Merge branch '1.0' into documentation/web3-modules

Commit:
typos and formatting improved in the web3-modules documentation

Commit:
parameters types updated
